Better bulk adds for List.

jamesqo opened the pull request:
This PR optimizes the `List(IEnumerable<T>)` .ctor and `AddRange` method for lazy enumerables.

- For the constructor, we were previously looping through the sequence and calling `Add` on each item. This is not very efficient since `Add` is not inlined. Instead, we can have a tighter loop where we essentially write `Add` inline and only update fields when we run out of space.

- For `InsertRange`, we check if we are adding at the end of the list first. This is a common case, since `AddRange` just forwards to `InsertRange`. If we are, then we call `AddEnumerable` instead of `Insert` in a loop.

I ran tests on corefx's System.Collections and everything passes.

Note: I can't test the perf impact these changes have at the moment because of #8266. If/when I am able to, however, I will update this PR.

cc @stephentoub @jkotas 

Related: https://github.com/dotnet/corefx/pull/13942#issuecomment-262844479

jamesqo commented:
cc @vancem

jamesqo commented:
@jkotas @stephentoub I managed to workaround the issue I was having. Here are the performance results I got: https://gist.github.com/jamesqo/5ac880d3e46d719c99ec388128ac50a6 Test code: https://github.com/jamesqo/Dotnet/blob/57417b93ed0290444b461b8e7dea2082b441b149/Program.cs#L49

They are not universal, but there is generally about a 10% improvement. I think all of the allocations involved may be dominating the results, and that's why the results are not always consistent.

justinvp commented:
Won't this introduce a behavior change?

Consider an enumerator that reads lines from a file or from the network that may throw from `MoveNext()`.

Or a contrived example:

```c#
public static IEnumerable<int> GetItems()
{
    yield return 1;
    yield return 2;
    yield return 3;
    yield return 4;
    throw new Exception();
}
```

Before this change, each item successfully enumerated before `MoveNext()` throws will be added to the list and `_size` will reflect the number of items successfully added.

After this change, those initial items will be inserted in the array, but then `MoveNext()` throws, so `_size` won't be updated to reflect the items that were successfully inserted.

jamesqo commented:
@justinvp, good point. I think I can wrap things in a try..catch or try..finally to ensure things are always updated.

jamesqo commented:
@justinvp Thanks for spotting that, I've fixed the problem and used try..finally.

@jkotas, @stephentoub Not sure what happened before, but I re-ran the perf test after fixing the issue. The results seem more consistent this time, improvements for larger collections are about 15-20%. I've updated the gist.

jamesqo commented:
@jkotas Sorry, can you please review when you have time? Thanks.

jkotas commented:
@jamesqo I am leaving tweaks of convenience methods like this one to [area maintainers](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/issue-guide.md) to deal with.

@Priya91 @ianhays PTLA

stephentoub commented on src/mscorlib/src/System/Collections/Generic/List.cs:
```diff
@@ -1031,6 +1028,45 @@ public void Sort(IComparer<T> comparer)
+                {
+                    // Make a final update to _size and _version after we've finished adding.
+                    _size = size;
+                    _version++;
```
These attempted optimizations around _size, _items, _version, etc. break code like:
```C#
var list = new List<int>();
list.AddRange(Enumerable.Range(0, 5).Select(i => { list.Add(i); return i; }));
```

jamesqo replied:
@stephentoub Interesting. I don't think it's going to be very common to modify the list in a selector, but seeing it in an invalid state/having items overwritten is pushing it.

I can remove the optimizations around `_size`, and increment `_version` before we do any of the actual copying. AFAICS caching `_items` is harmless though because the field is always up-to-date, it's just the local variable we're updating.

stephentoub replied:
> AFAICS caching _items is harmless though

How so?  If the iterator adds an element that causes the list to grow, ```_items``` will end up pointing to a different array than ```items``` points to, and AddRange will start storing the values into a garbage array.

jamesqo replied:
@stephentoub `EnsureCapacity` always updates `_items`, and we always update the local variable via `items = _items` after that. Unless an exception of some sort is thrown between those lines, in which case we stop copying anyways, the two are always equal.

stephentoub replied:
> EnsureCapacity always updates _items, and we always update the local variable via items = _items after that

But those lines aren't relevant to the case I'm talking about.  If the call to MoveNext increases the capacity of the list, items will now be different from _items, we won't enter that if block (as we will already have enough capacity since it was just resized), and we'll store the items into the wrong array.

jamesqo replied:
OK. I fixed that, and created a new gist: https://gist.github.com/jamesqo/a94ddead33dbf63118d6222980aabfc7 `AddRange` is still significantly faster. The method is basically an inline version of `Add` in a loop now.

jamesqo replied:
@stephentoub `Add` is not inlined. That is why the numbers show improvements.

jkotas replied:
cc @AndyAyersMS 

Manual inlining is double edged sword. You can get a small improvement in a lot of places by manually inlining and always show there is a performance improvement in microbenchmark. However if you have done manual inlining enough times, the real world workloads would likely get slower because of the code size increase from manual inlining would overwhelm the benefits.

We may want to be more data driven for manual inlining - only do it if there is real-world workload where it makes a dent, and leave profile guided optimizations to take care of the rest.

jamesqo replied:
> However if you have done manual inlining enough times, the real world workloads would likely get slower because of the code size increase from manual inlining would overwhelm the benefits.

I think it makes sense in this case. This in a loop and it's part of the bottleneck. It's also frequently used (directly, or indirectly by Linq).

> We may want to be more data driven for manual inlining - only do it if there is real-world workload where it makes a dent,

It just came up in corefx in a frequently-used method. See the PR I linked to: https://github.com/dotnet/corefx/pull/13942#issuecomment-262844479

AndyAyersMS replied:
@jamesqo Did you happen look into why Add is not getting inlined? It would be worth understanding that since at least from source it looks small enough that it should have been considered.

Also would be nice to see before/after disassembly in case the jit is just missing out on some obvious optimization.

jamesqo commented:
Updated the gist. There is still a 10% improvement.

jamesqo commented:
Test OSX x64 Checked Build and Test

justinvp commented on src/mscorlib/src/System/Collections/Generic/List.cs:
```diff
@@ -1031,6 +1028,31 @@ public void Sort(IComparer<T> comparer)
+
+            using (IEnumerator<T> en = enumerable.GetEnumerator())
+            {
+                _version++; // Even if the enumerable has no items, we can update _version.
```
`AddRange` already increments `_version`. Can we remove the increment here?

The other caller of this method is the .ctor, which shouldn't need the `_version` increment.

jamesqo replied:
@justinvp It is just in case if this is called by another method later on, or e.g. the increment to `_version` is accidentally removed in `InsertRange`. It is just one write anyway, so I don't think it should cost much relative to all the other stuff going on in the method.

jamesqo commented:
@ianhays, @Priya91, @stephentoub Sorry, do you think you could merge these changes when you have time? Thanks.
